src/biogeochem/CNBalanceCheckMod.F90

```diff
@@ -348,6 +524,171 @@ subroutine NBalanceCheck(this, bounds, num_soilc, filter_soilc, &
+         carr = som_n_leached(bounds%begc:bounds%endc), &
+         garr = som_n_leached_grc(bounds%begg:bounds%endg), &
+         c2l_scale_type = 'unity', &
+         l2g_scale_type = 'unity')
+      if (.not. use_nitrif_denitrif) then
+         call c2g( bounds = bounds, &
+            carr = sminn_leached(bounds%begc:bounds%endc), &
+            garr = sminn_leached_grc(bounds%begg:bounds%endg), &
+            c2l_scale_type = 'unity', &
+            l2g_scale_type = 'unity')
+      else
+         call c2g( bounds = bounds, &
+            carr = f_n2o_nit(bounds%begc:bounds%endc), &
+            garr = f_n2o_nit_grc(bounds%begg:bounds%endg), &
+            c2l_scale_type = 'unity', &
+            l2g_scale_type = 'unity')
+         call c2g( bounds = bounds, &
+            carr = smin_no3_leached(bounds%begc:bounds%endc), &
+            garr = smin_no3_leached_grc(bounds%begg:bounds%endg), &
+            c2l_scale_type = 'unity', &
+            l2g_scale_type = 'unity')
+         call c2g( bounds = bounds, &
+            carr = smin_no3_runoff(bounds%begc:bounds%endc), &
+            garr = smin_no3_runoff_grc(bounds%begg:bounds%endg), &
+            c2l_scale_type = 'unity', &
+            l2g_scale_type = 'unity')
+      end if
+
+      err_found = .false.
+      do g = bounds%begg, bounds%endg
+         ! calculate the total gridcell-level nitrogen storage, for mass conservation check
+         ! slevis: Not including seedn_grc in grc_begnb and grc_endnb because
+         ! seedn_grc equals
+         ! -1 * (dwt_seedn_to_leaf_grc(g) + dwt_seedn_to_deadstem_grc(g))
+         ! and we account for the latter fluxes as inputs below; the same
+         ! fluxes have entered the pools earlier in the timestep. For true
+         ! conservation we would need to add a flux out of nfix into seed.
+         grc_endnb(g) = totgrcn(g) + tot_woodprod_grc(g) + cropprod1_grc(g)
+
+         ! calculate total gridcell-level inputs
+         grc_ninputs(g) = forc_ndep(g) + nfix_to_sminn_grc(g) + &
+                          supplement_to_sminn_grc(g) + &
+                          dwt_seedn_to_leaf_grc(g) + &
+                          dwt_seedn_to_deadstem_grc(g)
+
+         if (use_fun) then
+            grc_ninputs(g) = grc_ninputs(g) + ffix_to_sminn_grc(g)
+         endif
+
+         if (use_crop) then
+            grc_ninputs(g) = grc_ninputs(g) + fert_to_sminn_grc(g) + soyfixn_to_sminn_grc(g)
+         end if
+
+         ! calculate total gridcell-level outputs
+         grc_noutputs(g) = denit_grc(g) + grc_fire_nloss(g) + &
+                           dwt_conv_nflux_grc(g) + product_loss_grc(g) - &
+                           som_n_leached_grc(g)
+
+         if (.not. use_nitrif_denitrif) then
+            grc_noutputs(g) = grc_noutputs(g) + sminn_leached_grc(g)
+         else
+            grc_noutputs(g) = grc_noutputs(g) + f_n2o_nit_grc(g) + &
+                              smin_no3_leached_grc(g) + smin_no3_runoff_grc(g)
+         end if
```

Comment:
- [ ] There is a lot of duplication in logic between the setting of these gridcell-level variables and the settings of the equivalent column-level variables above. Can you please remove this duplication by just averaging `col_ninputs` and `col_noutputs` to gridcell level? From a quick glance, it looks like all of the terms in `col_ninputs` and `col_noutputs` are also included in the gridcell level terms, but if there are differences, you could save an intermediate version of the column-level terms after adding up all terms that apply to both the col and gridcell level, then add any terms that *just* apply to the column level. Similarly, you could add any terms that *just* apply to gridcell level later.

By the way: I thought about suggesting something similar for carbon, but I actually like that, for carbon, you used the existing nbp term.

Reply:
Done.